Tweak E0277 output when a candidate is available

estebank opened the pull request:
*Follow up to #132086.*

Go from

```
error[E0277]: the trait bound `Then<Ignored<chumsky::combinator::Filter<chumsky::primitive::Any<&str, chumsky::extra::Full<EmptyErr, (), ()>>, {closure@src/main.rs:9:17: 9:27}>, char>, chumsky::combinator::Map<impl CSTParser<'a, O>, O, {closure@src/main.rs:11:24: 11:27}>, (), (), chumsky::extra::Full<EmptyErr, (), ()>>: CSTParser<'a>` is not satisfied
 --> src/main.rs:7:50
  |
7 | fn leaf<'a, O>(parser: impl CSTParser<'a, O>) -> impl CSTParser<'a, ()> {
  |                                                  ^^^^^^^^^^^^^^^^^^^^^^ the trait `chumsky::private::ParserSealed<'_, &str, (), chumsky::extra::Full<EmptyErr, (), ()>>` is not implemented for `Then<Ignored<Filter<Any<&str, ...>, ...>, ...>, ..., ..., ..., ...>`, which is required by `Then<Ignored<chumsky::combinator::Filter<chumsky::primitive::Any<&str, chumsky::extra::Full<EmptyErr, (), ()>>, {closure@src/main.rs:9:17: 9:27}>, char>, chumsky::combinator::Map<impl CSTParser<'a, O>, O, {closure@src/main.rs:11:24: 11:27}>, (), (), chumsky::extra::Full<EmptyErr, (), ()>>: CSTParser<'a>`
  |
  = help: the trait `chumsky::private::ParserSealed<'_, &'a str, ((), ()), chumsky::extra::Full<EmptyErr, (), ()>>` is implemented for `Then<Ignored<chumsky::combinator::Filter<chumsky::primitive::Any<&str, chumsky::extra::Full<EmptyErr, (), ()>>, {closure@src/main.rs:9:17: 9:27}>, char>, chumsky::combinator::Map<impl CSTParser<'a, O>, O, {closure@src/main.rs:11:24: 11:27}>, (), (), chumsky::extra::Full<EmptyErr, (), ()>>`
  = help: for that trait implementation, expected `((), ())`, found `()`
  = note: required for `Then<Ignored<Filter<Any<&str, ...>, ...>, ...>, ..., ..., ..., ...>` to implement `Parser<'_, &str, ()>`
note: required for `Then<Ignored<Filter<Any<&str, ...>, ...>, ...>, ..., ..., ..., ...>` to implement `CSTParser<'a>`
 --> src/main.rs:5:16
  |
5 | impl<'a, O, T> CSTParser<'a, O> for T where T: Parser<'a, &'a str, O> {}
  |                ^^^^^^^^^^^^^^^^     ^          ---------------------- unsatisfied trait bound introduced here
  = note: the full name for the type has been written to '/home/gh-estebank/longlong/target/debug/deps/longlong-0008f9a4f2023b08.long-type-13239977239800463552.txt'
  = note: consider using `--verbose` to print the full type name to the console
  = note: the full name for the type has been written to '/home/gh-estebank/longlong/target/debug/deps/longlong-0008f9a4f2023b08.long-type-13239977239800463552.txt'
  = note: consider using `--verbose` to print the full type name to the console
```

to

```
error[E0277]: the trait bound `Then<Ignored<chumsky::combinator::Filter<chumsky::primitive::Any<&str, chumsky::extra::Full<EmptyErr, (), ()>>, {closure@src/main.rs:9:17: 9:27}>, char>, chumsky::combinator::Map<impl CSTParser<'a, O>, O, {closure@src/main.rs:11:24: 11:27}>, (), (), chumsky::extra::Full<EmptyErr, (), ()>>: CSTParser<'a>` is not satisfied
  --> src/main.rs:7:50
   |
7  | fn leaf<'a, O>(parser: impl CSTParser<'a, O>) -> impl CSTParser<'a, ()> {
   |                                                  ^^^^^^^^^^^^^^^^^^^^^^ unsatisfied trait bound
...
11 |     ws.then(parser.map(|_| ()))
   |     --------------------------- return type was inferred to be `Then<Ignored<..., ...>, ..., ..., ..., ...>` here
   |
   = help: the trait `ParserSealed<'_, &_, (), Full<_, _, _>>` is not implemented for `Then<Ignored<..., ...>, ..., ..., ..., ...>`
           but trait `ParserSealed<'_, &'a _, ((), ()), Full<_, _, _>>` is implemented for it
   = help: for that trait implementation, expected `((), ())`, found `()`
   = note: required for `Then<Ignored<..., ...>, ..., ..., ..., ...>` to implement `Parser<'_, &str, ()>`
note: required for `Then<Ignored<..., ...>, ..., ..., ..., ...>` to implement `CSTParser<'a>`
  --> src/main.rs:5:16
   |
5  | impl<'a, O, T> CSTParser<'a, O> for T where T: Parser<'a, &'a str, O> {}
   |                ^^^^^^^^^^^^^^^^     ^          ---------------------- unsatisfied trait bound introduced here
   = note: the full name for the type has been written to '/home/gh-estebank/longlong/target/debug/deps/longlong-df9d52be87eada65.long-type-1337037744507305372.txt'
   = note: consider using `--verbose` to print the full type name to the console
```

* Remove redundant wording
* Introduce trait diff highlighting logic and use it
* Fix incorrect "long type written to path" logic (can be split off)
* Point at tail expression in more cases in E0277
* Avoid long primary span labels in E0277 by moving them to a `help`

Fix #132013.

There are individual commits that can be their own PR. If the review load is too big, happy to split them off.

estebank committed:
Add trait diff highlighting logic and use it in E0277

When a trait is not implemented for a type, but there *is* an `impl`
for another type or different trait params, we format the output to
use highlighting in the same way that E0308 does for types.

The logic accounts for 3 cases:
- When both the type and trait in the expected predicate and the candidate are different
- When only the types are different
- When only the trait generic params are different

For each case, we use slightly different formatting and wording.

estebank committed:
Trim output of E0277 in some cases

Remove default note for "trait is not implemented" in favor of the
more colorful diff output from the previous commit. Removes
duplicated output.

estebank committed:
Fix type shortening writing to file

Make sure that we append to the file for long ty paths. Do not write the same type more than once. Shorten the calculated width a bit.

estebank committed:
Use `short_ty_string`

estebank committed:
Point at tail expression on rpit E0277

```
error[E0277]: the trait bound `{gen block@$DIR/gen_block_is_coro.rs:7:5: 7:8}: Coroutine` is not satisfied
  --> $DIR/gen_block_is_coro.rs:6:13
   |
LL | fn foo() -> impl Coroutine<Yield = u32, Return = ()> {
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Coroutine` is not implemented for `{gen block@$DIR/gen_block_is_coro.rs:7:5: 7:8}`
LL |     gen { yield 42 }
   |     ---------------- return type was inferred to be `{gen block@$DIR/gen_block_is_coro.rs:7:5: 7:8}` here
```

The secondary span label is new.

estebank committed:
On long E0277 primary span label, move it to a `help`

Long span labels don't read well.

estebank commented on tests/ui/impl-trait/diagnostics/highlight-difference-between-expected-trait-and-found-trait.svg:
This is the main test we want to look at, but some of the changes (like tweaking the printing of diffed types and of the long type file) aren't present here.

The output for the repro:

![Screenshot 2024-10-25 at 11 10 14 AM](https://github.com/user-attachments/assets/5675a5ce-59e2-4c85-b8ff-068fbb66671a)

estebank replied:
After all of the other changes, I am tempted at removing the "for that trait implementation, expected..." `help`, as it makes it more explicit, but is already implicitly present in the line above.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11523223329/job/32080820503) [(plain)](https://github.com/rust-lang/rust/commit/3a764ba8dd8102e271a1c1be2422be3e23f46018/checks/32080820503/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:c32c805632780b5c1de330e3f44561b336c2efe163bc0990acb392390157a8e1d9f855d75914a239aa40c49d77f4a837247d05d2f8d46f554b98e1f46712a3e3:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
diff of stderr:

30   --> $DIR/coroutine-not-future.rs:39:21
31    |
32 LL |     takes_coroutine(async {});
-    |     --------------- ^^^^^^^^ unsatisfied trait bound
+    |     --------------- ^^^^^^^^ the trait `Coroutine<_>` is not implemented for `{async block@$DIR/coroutine-not-future.rs:39:21: 39:26}`
35    |     required by a bound introduced by this call
36    |


-    = help: the trait `Coroutine<_>` is not implemented for `{async block@$DIR/coroutine-not-future.rs:39:21: 39:26}`
38 note: required by a bound in `takes_coroutine`
40    |


The actual stderr differed from the expected stderr.
The actual stderr differed from the expected stderr.
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args async-await/coroutine-not-future.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/async-await/coroutine-not-future.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/async-await/coroutine-not-future" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/async-await/coroutine-not-future/auxiliary" "--edition=2018"
--- stderr -------------------------------
--- stderr -------------------------------
error[E0277]: the trait bound `impl Future<Output = ()>: Coroutine<_>` is not satisfied
   |
   |
LL |     takes_coroutine(async_fn());
   |     --------------- ^^^^^^^^^^ the trait `Coroutine<_>` is not implemented for `impl Future<Output = ()>`
   |     required by a bound introduced by this call
   |
note: required by a bound in `takes_coroutine`
  --> /checkout/tests/ui/async-await/coroutine-not-future.rs:19:39
  --> /checkout/tests/ui/async-await/coroutine-not-future.rs:19:39
   |
LL | fn takes_coroutine<ResumeTy>(_g: impl Coroutine<ResumeTy, Yield = (), Return = ()>) {}
   |                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `takes_coroutine`

error[E0277]: the trait bound `impl Future<Output = ()>: Coroutine<_>` is not satisfied
   |
LL |     takes_coroutine(returns_async_block());
   |     --------------- ^^^^^^^^^^^^^^^^^^^^^ the trait `Coroutine<_>` is not implemented for `impl Future<Output = ()>`
   |     |
   |     |
   |     required by a bound introduced by this call
   |
note: required by a bound in `takes_coroutine`
  --> /checkout/tests/ui/async-await/coroutine-not-future.rs:19:39
   |
LL | fn takes_coroutine<ResumeTy>(_g: impl Coroutine<ResumeTy, Yield = (), Return = ()>) {}
   |                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `takes_coroutine`

error[E0277]: the trait bound `{async block@/checkout/tests/ui/async-await/coroutine-not-future.rs:39:21: 39:26}: Coroutine<_>` is not satisfied
   |
   |
LL |     takes_coroutine(async {});
   |     --------------- ^^^^^^^^ the trait `Coroutine<_>` is not implemented for `{async block@/checkout/tests/ui/async-await/coroutine-not-future.rs:39:21: 39:26}`
   |     required by a bound introduced by this call
   |
note: required by a bound in `takes_coroutine`
  --> /checkout/tests/ui/async-await/coroutine-not-future.rs:19:39
  --> /checkout/tests/ui/async-await/coroutine-not-future.rs:19:39
   |
LL | fn takes_coroutine<ResumeTy>(_g: impl Coroutine<ResumeTy, Yield = (), Return = ()>) {}
   |                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `takes_coroutine`

error[E0277]: `impl Coroutine<Yield = (), Return = ()>` is not a future
   |
LL |     takes_future(returns_coroutine());
LL |     takes_future(returns_coroutine());
   |     ------------ ^^^^^^^^^^^^^^^^^^^ `impl Coroutine<Yield = (), Return = ()>` is not a future
   |     required by a bound introduced by this call
   |
   |
   = help: the trait `Future` is not implemented for `impl Coroutine<Yield = (), Return = ()>`
note: required by a bound in `takes_future`
   |
   |
LL | fn takes_future(_f: impl Future<Output = ()>) {}
   |                          ^^^^^^^^^^^^^^^^^^^ required by this bound in `takes_future`

error[E0277]: `{coroutine@/checkout/tests/ui/async-await/coroutine-not-future.rs:47:9: 47:14}` is not a future
   |
LL |       takes_future(
   |       ------------ required by a bound introduced by this call
LL |           #[coroutine]
LL |           #[coroutine]
LL | /         |ctx| {
LL | |             //~^ ERROR is not a future
LL | |             ctx = yield ();
LL | |         },
   | |_________^ `{coroutine@/checkout/tests/ui/async-await/coroutine-not-future.rs:47:9: 47:14}` is not a future
   |
   = help: the trait `Future` is not implemented for `{coroutine@/checkout/tests/ui/async-await/coroutine-not-future.rs:47:9: 47:14}`
note: required by a bound in `takes_future`
   |
   |
LL | fn takes_future(_f: impl Future<Output = ()>) {}
   |                          ^^^^^^^^^^^^^^^^^^^ required by this bound in `takes_future`
error: aborting due to 5 previous errors

For more information about this error, try `rustc --explain E0277`.
------------------------------------------
------------------------------------------


---- [ui] tests/ui/coroutine/clone-impl.rs stdout ----
Saved the actual stderr to "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/coroutine/clone-impl/clone-impl.stderr"
diff of stderr:

25    |     ------- within this `{coroutine@$DIR/clone-impl.rs:43:5: 43:12}`
26 ...
27 LL |     check_copy(&gen_clone_0);
-    |     ^^^^^^^^^^^^^^^^^^^^^^^^ unsatisfied trait bound
+    |     ^^^^^^^^^^^^^^^^^^^^^^^^ within `{coroutine@$DIR/clone-impl.rs:43:5: 43:12}`, the trait `Copy` is not implemented for `Vec<char>`
29    |
-    = help: within `{coroutine@$DIR/clone-impl.rs:43:5: 43:12}`, the trait `Copy` is not implemented for `Vec<char>`
31 note: coroutine does not implement `Copy` as this value is used across a yield
33    |


68    |     ------- within this `{coroutine@$DIR/clone-impl.rs:58:5: 58:12}`
69 ...
70 LL |     check_copy(&gen_clone_1);
-    |     ^^^^^^^^^^^^^^^^^^^^^^^^ unsatisfied trait bound
+    |     ^^^^^^^^^^^^^^^^^^^^^^^^ within `{coroutine@$DIR/clone-impl.rs:58:5: 58:12}`, the trait `Copy` is not implemented for `Vec<char>`
72    |
-    = help: within `{coroutine@$DIR/clone-impl.rs:58:5: 58:12}`, the trait `Copy` is not implemented for `Vec<char>`
74 note: coroutine does not implement `Copy` as this value is used across a yield
76    |


117    |     ------- within this `{coroutine@$DIR/clone-impl.rs:79:5: 79:12}`
118 ...
119 LL |     check_clone(&gen_non_clone);
-    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^ unsatisfied trait bound
+    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^ within `{coroutine@$DIR/clone-impl.rs:79:5: 79:12}`, the trait `Clone` is not implemented for `NonClone`
121    |
-    = help: within `{coroutine@$DIR/clone-impl.rs:79:5: 79:12}`, the trait `Clone` is not implemented for `NonClone`
123 note: captured value does not implement `Clone`
125    |


The actual stderr differed from the expected stderr.
The actual stderr differed from the expected stderr.
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args coroutine/clone-impl.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/coroutine/clone-impl.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/coroutine/clone-impl" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/coroutine/clone-impl/auxiliary"
--- stderr -------------------------------
--- stderr -------------------------------
error[E0277]: the trait bound `Vec<u32>: Copy` is not satisfied in `{coroutine@/checkout/tests/ui/coroutine/clone-impl.rs:43:5: 43:12}`
   |
LL |     move || {
LL |     move || {
   |     ------- within this `{coroutine@/checkout/tests/ui/coroutine/clone-impl.rs:43:5: 43:12}`
...
LL |     check_copy(&gen_clone_0);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^ within `{coroutine@/checkout/tests/ui/coroutine/clone-impl.rs:43:5: 43:12}`, the trait `Copy` is not implemented for `Vec<u32>`
note: captured value does not implement `Copy`
  --> /checkout/tests/ui/coroutine/clone-impl.rs:47:14
   |
LL |         drop(clonable_0);
LL |         drop(clonable_0);
   |              ^^^^^^^^^^ has type `Vec<u32>` which does not implement `Copy`
note: required by a bound in `check_copy`
   |
   |
LL | fn check_copy<T: Copy>(_x: &T) {}
   |                  ^^^^ required by this bound in `check_copy`

error[E0277]: the trait bound `Vec<char>: Copy` is not satisfied in `{coroutine@/checkout/tests/ui/coroutine/clone-impl.rs:43:5: 43:12}`
   |
LL |     move || {
LL |     move || {
   |     ------- within this `{coroutine@/checkout/tests/ui/coroutine/clone-impl.rs:43:5: 43:12}`
...
LL |     check_copy(&gen_clone_0);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^ within `{coroutine@/checkout/tests/ui/coroutine/clone-impl.rs:43:5: 43:12}`, the trait `Copy` is not implemented for `Vec<char>`
note: coroutine does not implement `Copy` as this value is used across a yield
  --> /checkout/tests/ui/coroutine/clone-impl.rs:45:9
   |
   |
LL |         let v = vec!['a'];
   |             - has type `Vec<char>` which does not implement `Copy`
LL |         yield;
   |         ^^^^^ yield occurs here, with `v` maybe used later
note: required by a bound in `check_copy`
   |
   |
LL | fn check_copy<T: Copy>(_x: &T) {}
   |                  ^^^^ required by this bound in `check_copy`

error[E0277]: the trait bound `Vec<u32>: Copy` is not satisfied in `{coroutine@/checkout/tests/ui/coroutine/clone-impl.rs:58:5: 58:12}`
   |
LL |     move || {
LL |     move || {
   |     ------- within this `{coroutine@/checkout/tests/ui/coroutine/clone-impl.rs:58:5: 58:12}`
...
LL |     check_copy(&gen_clone_1);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^ within `{coroutine@/checkout/tests/ui/coroutine/clone-impl.rs:58:5: 58:12}`, the trait `Copy` is not implemented for `Vec<u32>`
note: captured value does not implement `Copy`
  --> /checkout/tests/ui/coroutine/clone-impl.rs:68:14
   |
LL |         drop(clonable_1);
LL |         drop(clonable_1);
   |              ^^^^^^^^^^ has type `Vec<u32>` which does not implement `Copy`
note: required by a bound in `check_copy`
   |
   |
LL | fn check_copy<T: Copy>(_x: &T) {}
   |                  ^^^^ required by this bound in `check_copy`

error[E0277]: the trait bound `Vec<char>: Copy` is not satisfied in `{coroutine@/checkout/tests/ui/coroutine/clone-impl.rs:58:5: 58:12}`
   |
LL |     move || {
LL |     move || {
   |     ------- within this `{coroutine@/checkout/tests/ui/coroutine/clone-impl.rs:58:5: 58:12}`
...
LL |     check_copy(&gen_clone_1);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^ within `{coroutine@/checkout/tests/ui/coroutine/clone-impl.rs:58:5: 58:12}`, the trait `Copy` is not implemented for `Vec<char>`
note: coroutine does not implement `Copy` as this value is used across a yield
  --> /checkout/tests/ui/coroutine/clone-impl.rs:64:9
   |
   |
LL |         let v = vec!['a'];
   |             - has type `Vec<char>` which does not implement `Copy`
LL |         yield;
LL |         yield;
   |         ^^^^^ yield occurs here, with `v` maybe used later
note: required by a bound in `check_copy`
   |
   |
LL | fn check_copy<T: Copy>(_x: &T) {}
   |                  ^^^^ required by this bound in `check_copy`

error[E0277]: the trait bound `NonClone: Copy` is not satisfied in `{coroutine@/checkout/tests/ui/coroutine/clone-impl.rs:79:5: 79:12}`
   |
LL |     move || {
LL |     move || {
   |     ------- within this `{coroutine@/checkout/tests/ui/coroutine/clone-impl.rs:79:5: 79:12}`
...
LL |     check_copy(&gen_non_clone);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^ within `{coroutine@/checkout/tests/ui/coroutine/clone-impl.rs:79:5: 79:12}`, the trait `Copy` is not implemented for `NonClone`
note: captured value does not implement `Copy`
  --> /checkout/tests/ui/coroutine/clone-impl.rs:81:14
   |
LL |         drop(non_clonable);
LL |         drop(non_clonable);
   |              ^^^^^^^^^^^^ has type `NonClone` which does not implement `Copy`
note: required by a bound in `check_copy`
  --> /checkout/tests/ui/coroutine/clone-impl.rs:89:18
   |
LL | fn check_copy<T: Copy>(_x: &T) {}
   |                  ^^^^ required by this bound in `check_copy`
help: consider annotating `NonClone` with `#[derive(Copy)]`
LL + #[derive(Copy)]
LL | struct NonClone;
   |


error[E0277]: the trait bound `NonClone: Clone` is not satisfied in `{coroutine@/checkout/tests/ui/coroutine/clone-impl.rs:79:5: 79:12}`
   |
LL |     move || {
LL |     move || {
   |     ------- within this `{coroutine@/checkout/tests/ui/coroutine/clone-impl.rs:79:5: 79:12}`
...
LL |     check_clone(&gen_non_clone);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^ within `{coroutine@/checkout/tests/ui/coroutine/clone-impl.rs:79:5: 79:12}`, the trait `Clone` is not implemented for `NonClone`
note: captured value does not implement `Clone`
  --> /checkout/tests/ui/coroutine/clone-impl.rs:81:14
   |
LL |         drop(non_clonable);
LL |         drop(non_clonable);
   |              ^^^^^^^^^^^^ has type `NonClone` which does not implement `Clone`
note: required by a bound in `check_clone`
  --> /checkout/tests/ui/coroutine/clone-impl.rs:90:19
   |
LL | fn check_clone<T: Clone>(_x: &T) {}
   |                   ^^^^^ required by this bound in `check_clone`
help: consider annotating `NonClone` with `#[derive(Clone)]`
LL + #[derive(Clone)]
LL | struct NonClone;
   |

---
---- [ui] tests/ui/coroutine/issue-105084.rs stdout ----
Saved the actual stderr to "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/coroutine/issue-105084/issue-105084.stderr"
diff of stderr:

29    |     -- within this `{coroutine@$DIR/issue-105084.rs:15:5: 15:7}`
30 ...
31 LL |     let mut h = copy(g);
-    |                 ^^^^^^^ unsatisfied trait bound
+    |                 ^^^^^^^ within `{coroutine@$DIR/issue-105084.rs:15:5: 15:7}`, the trait `Copy` is not implemented for `Box<(i32, ())>`
33    |
-    = help: within `{coroutine@$DIR/issue-105084.rs:15:5: 15:7}`, the trait `Copy` is not implemented for `Box<(i32, ())>`
35 note: coroutine does not implement `Copy` as this value is used across a yield
37    |


The actual stderr differed from the expected stderr.
The actual stderr differed from the expected stderr.
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args coroutine/issue-105084.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/coroutine/issue-105084.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/coroutine/issue-105084" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/coroutine/issue-105084/auxiliary"
--- stderr -------------------------------
error[E0382]: borrow of moved value: `g`
##[error]  --> /checkout/tests/ui/coroutine/issue-105084.rs:38:14
   |
   |
LL |     let mut g = #[coroutine]
   |         ----- move occurs because `g` has type `{coroutine@/checkout/tests/ui/coroutine/issue-105084.rs:15:5: 15:7}`, which does not implement the `Copy` trait
...
LL |     let mut h = copy(g);
...
...
LL |     Pin::new(&mut g).resume(());
   |
note: consider changing this parameter type in function `copy` to borrow instead if owning the value isn't necessary
  --> /checkout/tests/ui/coroutine/issue-105084.rs:9:21
   |
   |
LL | fn copy<T: Copy>(x: T) -> T {
   |    ----             ^ this parameter takes ownership of the value
   |    in this function
help: consider cloning the value if the performance cost is acceptable
   |
   |
LL |     let mut h = copy(g.clone());


error[E0277]: the trait bound `Box<(i32, ())>: Copy` is not satisfied in `{coroutine@/checkout/tests/ui/coroutine/issue-105084.rs:15:5: 15:7}`
   |
LL |     || {
LL |     || {
   |     -- within this `{coroutine@/checkout/tests/ui/coroutine/issue-105084.rs:15:5: 15:7}`
...
LL |     let mut h = copy(g);
   |                 ^^^^^^^ within `{coroutine@/checkout/tests/ui/coroutine/issue-105084.rs:15:5: 15:7}`, the trait `Copy` is not implemented for `Box<(i32, ())>`
note: coroutine does not implement `Copy` as this value is used across a yield
  --> /checkout/tests/ui/coroutine/issue-105084.rs:22:22
   |
   |
LL |         Box::new((5, yield));
   |         |            |
   |         |            |
   |         |            yield occurs here, with `Box::new((5, yield))` maybe used later
   |         has type `Box<(i32, ())>` which does not implement `Copy`
  --> /checkout/tests/ui/coroutine/issue-105084.rs:9:12
   |
   |
LL | fn copy<T: Copy>(x: T) -> T {

error: aborting due to 2 previous errors

Some errors have detailed explanations: E0277, E0382.
---
diff of stderr:

40   --> $DIR/issue-84973-blacklist.rs:17:26
41    |
42 LL |     f_unpin(#[coroutine] static || { yield; });
-    |     -------              ^^^^^^^^^^^^^^^^^^^^ unsatisfied trait bound
+    |     -------              ^^^^^^^^^^^^^^^^^^^^ the trait `Unpin` is not implemented for `{static coroutine@$DIR/issue-84973-blacklist.rs:17:26: 17:35}`
45    |     required by a bound introduced by this call
46    |


-    = help: the trait `Unpin` is not implemented for `{static coroutine@$DIR/issue-84973-blacklist.rs:17:26: 17:35}`
48    = note: consider using the `pin!` macro
49            consider using `Box::pin` if you need to access the pinned value outside of the current scope
50 note: required by a bound in `f_unpin`

The actual stderr differed from the expected stderr.
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args suggestions/issue-84973-blacklist.rs`
To only update this specific test, also pass `--test-args suggestions/issue-84973-blacklist.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/suggestions/issue-84973-blacklist.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/suggestions/issue-84973-blacklist" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/suggestions/issue-84973-blacklist/auxiliary"
--- stderr -------------------------------
error[E0277]: the trait bound `String: Copy` is not satisfied
##[error]  --> /checkout/tests/ui/suggestions/issue-84973-blacklist.rs:15:12
   |
   |
LL |     f_copy("".to_string()); //~ ERROR: the trait bound `String: Copy` is not satisfied [E0277]
   |     |
   |     required by a bound introduced by this call
   |
note: required by a bound in `f_copy`
note: required by a bound in `f_copy`
  --> /checkout/tests/ui/suggestions/issue-84973-blacklist.rs:6:14
   |
LL | fn f_copy<T: Copy>(t: T) {}
   |              ^^^^ required by this bound in `f_copy`
help: consider removing this method call, as the receiver has type `&'static str` and `&'static str: Copy` trivially holds
   |
LL -     f_copy("".to_string()); //~ ERROR: the trait bound `String: Copy` is not satisfied [E0277]
LL +     f_copy(""); //~ ERROR: the trait bound `String: Copy` is not satisfied [E0277]

error[E0277]: the trait bound `S: Clone` is not satisfied
##[error]  --> /checkout/tests/ui/suggestions/issue-84973-blacklist.rs:16:13
   |
   |
LL |     f_clone(S); //~ ERROR: the trait bound `S: Clone` is not satisfied [E0277]
   |     ------- ^ the trait `Clone` is not implemented for `S`
   |     required by a bound introduced by this call
   |
note: required by a bound in `f_clone`
  --> /checkout/tests/ui/suggestions/issue-84973-blacklist.rs:7:15
  --> /checkout/tests/ui/suggestions/issue-84973-blacklist.rs:7:15
   |
LL | fn f_clone<T: Clone>(t: T) {}
   |               ^^^^^ required by this bound in `f_clone`
help: consider annotating `S` with `#[derive(Clone)]`
LL + #[derive(Clone)]
LL | struct S;
   |


error[E0277]: `{static coroutine@/checkout/tests/ui/suggestions/issue-84973-blacklist.rs:17:26: 17:35}` cannot be unpinned
   |
   |
LL |     f_unpin(#[coroutine] static || { yield; });
   |     -------              ^^^^^^^^^^^^^^^^^^^^ the trait `Unpin` is not implemented for `{static coroutine@/checkout/tests/ui/suggestions/issue-84973-blacklist.rs:17:26: 17:35}`
   |     required by a bound introduced by this call
   |
   = note: consider using the `pin!` macro
           consider using `Box::pin` if you need to access the pinned value outside of the current scope
           consider using `Box::pin` if you need to access the pinned value outside of the current scope
note: required by a bound in `f_unpin`
   |
   |
LL | fn f_unpin<T: Unpin>(t: T) {}
   |               ^^^^^ required by this bound in `f_unpin`
error[E0277]: the size for values of type `dyn Fn()` cannot be known at compilation time
##[error]  --> /checkout/tests/ui/suggestions/issue-84973-blacklist.rs:22:13
   |
   |
LL |     f_sized(*ref_cl);
   |     |
   |     required by a bound introduced by this call
   |
   = help: the trait `Sized` is not implemented for `dyn Fn()`
   = help: the trait `Sized` is not implemented for `dyn Fn()`
note: required by a bound in `f_sized`
  --> /checkout/tests/ui/suggestions/issue-84973-blacklist.rs:9:15
   |
LL | fn f_sized<T: Sized>(t: T) {}
   |               ^^^^^ required by this bound in `f_sized`
help: references are always `Sized`, even if they point to unsized data; consider not dereferencing the expression
   |
LL -     f_sized(*ref_cl);
LL +     f_sized(ref_cl);


error[E0277]: `Rc<{integer}>` cannot be sent between threads safely
   |
   |
LL |     f_send(rc); //~ ERROR: `Rc<{integer}>` cannot be sent between threads safely [E0277]
   |     ------ ^^ `Rc<{integer}>` cannot be sent between threads safely
   |     required by a bound introduced by this call
   |
   = help: the trait `Send` is not implemented for `Rc<{integer}>`
note: required by a bound in `f_send`
note: required by a bound in `f_send`
  --> /checkout/tests/ui/suggestions/issue-84973-blacklist.rs:10:14
   |
LL | fn f_send<T: Send>(t: T) {}
   |              ^^^^ required by this bound in `f_send`
error: aborting due to 5 previous errors

For more information about this error, try `rustc --explain E0277`.
------------------------------------------
```

</details>

compiler-errors commented:
I think splitting out the fourth commit, which touches almost every E0277, is probably the best action. Right now, it makes it practically impossible to critique any other change in this PR, because right now those practically just drown in the diff. This is especially true because the rest of the changes stacked on top have very little to do with each other, and imo could land in parallel.

estebank commented:
Split-off 4th commit to https://github.com/rust-lang/rust/pull/132157.

estebank commented on tests/ui/traits/negative-bounds/simple.stderr:
```diff
@@ -28,7 +28,8 @@ error[E0277]: the trait bound `Copyable: !Copy` is not satisfied
    |
-   = help: the trait `Copy` is implemented for `Copyable`
+   = help: the trait `Copy` is not implemented for `Copyable`
+           but trait `Copy` is implemented for it
```
Oh, that's unfortunate. Pre-existing, but unfortunate.

estebank replied:
fixed

estebank committed:
Account for negative bounds in E0277 note and suggestion

Do not suggest `#[derive(Copy)]` when we wanted a `!Copy` type.

Do not say "`Copy` is not implemented for `T` but `Copy` is".

Do not talk about `Trait` having no implementations when `!Trait` was desired.

Nadrieril commented:
r? compiler

bors commented:
:umbrella: The latest upstream changes (presumably #132326) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

davidtwco commented:
r=me after rebasing

estebank commented:
@bors r=davidtwco

matthiaskrgr commented:
@bors r- 
needs rebase I think

estebank committed:
fix tests

Paths in CI can be longer than in devs' machines.

estebank commented:
@matthiaskrgr it ended up being that the type name length of types that include paths (closures and coroutines) is longer in CI than in devs' machines because in CI they show the full path, and then change that customized output to avoid longer span labels triggered in CI and not in my environment :-/

@bors r=davidtwco

bors commented:
:pushpin: Commit 7e0d3ed9511f17097f9c88d849321267debd4a33 has been approved by `davidtwco`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=davidtwco 7e0d3ed9511f17097f9c88d849321267debd4a33 -->
<!-- homu: {"type":"Approved","sha":"7e0d3ed9511f17097f9c88d849321267debd4a33","approver":"davidtwco","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 7e0d3ed9511f17097f9c88d849321267debd4a33 with merge b3f75cc872cfd306860c3ad76a239e719015f855...
<!-- homu: {"type":"BuildStarted","head_sha":"7e0d3ed9511f17097f9c88d849321267debd4a33","merge_sha":"b3f75cc872cfd306860c3ad76a239e719015f855"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/11645568372/job/32430482701)
Approved by: davidtwco
Pushing b3f75cc872cfd306860c3ad76a239e719015f855 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"davidtwco","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/11645568372/job/32430482701"},"merge_sha":"b3f75cc872cfd306860c3ad76a239e719015f855"} -->

rust-timer commented:
Finished benchmarking commit (b3f75cc872cfd306860c3ad76a239e719015f855): [comparison URL](https://perf.rust-lang.org/compare.html?start=00ed73cdc09a6452cb58202d56a9211fb3c73031&end=b3f75cc872cfd306860c3ad76a239e719015f855&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=00ed73cdc09a6452cb58202d56a9211fb3c73031&end=b3f75cc872cfd306860c3ad76a239e719015f855&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 1.4%  | [1.4%, 1.4%]   | 1     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.1% | [-2.1%, -2.1%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=00ed73cdc09a6452cb58202d56a9211fb3c73031&end=b3f75cc872cfd306860c3ad76a239e719015f855&stat=max-rss)
<details>
<summary>Results (primary 4.1%, secondary 4.5%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 4.1% | [4.1%, 4.1%] | 1     |
| Regressions ❌ <br /> (secondary)  | 4.5% | [4.5%, 4.5%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 4.1% | [4.1%, 4.1%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=00ed73cdc09a6452cb58202d56a9211fb3c73031&end=b3f75cc872cfd306860c3ad76a239e719015f855&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=00ed73cdc09a6452cb58202d56a9211fb3c73031&end=b3f75cc872cfd306860c3ad76a239e719015f855&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 780.047s -> 781.635s (0.20%)
**Artifact size**: 335.36 MiB -> 335.39 MiB (0.01%)
